ToolBox/IQFeedDownloader/IQFeedDataDownloader.cs

```diff
@@ -0,0 +1,119 @@
+        /// <param name="resolution">Resolution of the data request</param>
+        /// <param name="startUtc">Start time of the data in UTC</param>
+        /// <param name="endUtc">End time of the data in UTC</param>
+        /// <returns>Enumerable of base data for this symbol</returns>
+        public IEnumerable<BaseData> Get(Symbol symbol, Resolution resolution, DateTime startUtc, DateTime endUtc)
+        {
+            if (symbol.ID.SecurityType != SecurityType.Equity)
+                throw new NotSupportedException("SecurityType not available: " + symbol.ID.SecurityType);
+
+            if (endUtc < startUtc)
+                throw new ArgumentException("The end date must be greater or equal than the start date.");
+
+            var dataType = resolution == Resolution.Tick ? typeof(Tick) : typeof(TradeBar);
+            var tickType = resolution == Resolution.Tick ? TickType.Quote : TickType.Trade;
+
+            var slices = _historyPort.ProcessHistoryRequests(
+                new HistoryRequest(
+                    startUtc,
+                    endUtc,
+                    dataType,
+                    symbol,
+                    resolution,
+                    SecurityExchangeHours.AlwaysOpen(TimeZones.NewYork),
+                    TimeZones.NewYork,
+                    resolution,
+                    true,
+                    false,
+                    DataNormalizationMode.Adjusted,
+                    tickType)).ToList();
+
+            switch (resolution)
+            {
+                case Resolution.Tick:
+                    var ticks = new List<Tick>();
+                    foreach (var slice in slices)
+                    {
+                        ticks.AddRange(slice.Ticks[symbol]);
+                    }
+                    return ticks;
+
+                case Resolution.Second:
+                case Resolution.Minute:
+                case Resolution.Hour:
+                case Resolution.Daily:
+                    var tradeBars = new List<TradeBar>();
+                    foreach (var slice in slices)
+                    {
+                        tradeBars.Add(slice.Bars[symbol]);
+                    }
+                    return tradeBars;
+
+                default:
+                    throw new NotSupportedException("Resolution not available: " + resolution);
+            }
+        }
+
+        /// <summary>
+        /// Aggregates a list of ticks at the requested resolution
+        /// </summary>
+        /// <param name="symbol"></param>
+        /// <param name="ticks"></param>
+        /// <param name="resolution"></param>
+        /// <returns></returns>
+        internal static IEnumerable<TradeBar> AggregateTicks(Symbol symbol, IEnumerable<Tick> ticks, TimeSpan resolution)
+        {
+            return (from t in ticks
+                    group t by t.Time.RoundDown(resolution) into g
+                    select new TradeBar
+                    {
```

Comment:
might want to explicitly set the `Period` property of trade bar to the provided `resolution` -- this will ensure the correct `EndTime` will be set. In LEAN, all data is time synced by the end time, so it turns out to be most important to get that time correct.

Reply:
Before develop this downloader, i work on iqfeed api, especially on their api examples. After a while i see actually lean already have almost all required codes for accomplish this task. Because i don't want to duplicate many lines of code, i do small changes current code and use them. For example used current HistoryPort implementation for request history from iqfeed's history socket port. After decide to support write all resolutions with a single command, just like HistoryPort i try to follow up current (similar) codes and check other donwloaders and see they are aggregating ticks without set EndTime property. I did not encounter any problems about that. But of course it should be examined in more detail.

Reply:
I think we should still verify the correctness of this to ensure the trade bars are creating with the correct end times as that's what LEAN uses for time synchronization.